File: app/src/main/java/org/sil/storyproducer/controller/pager/CircularViewPagerHandler.java
package org.sil.storyproducer.controller.pager;

import android.support.v4.view.ViewPager;

/**
 * Class that implements the ViewPager.OnPageChangeListener to give the view pager circular functionality
 */
public class CircularViewPagerHandler implements ViewPager.OnPageChangeListener {
    private ViewPager   mViewPager;
    private int         mCurrentPosition;
    private int         mScrollState;

    public CircularViewPagerHandler(final ViewPager viewPager) {
        mViewPager = viewPager;
    }

    @Override
    public void onPageSelected(final int position) {
        mCurrentPosition = position;
    }

    @Override
    public void onPageScrollStateChanged(final int state) {
        if (state == ViewPager.SCROLL_STATE_IDLE && mScrollState != ViewPager.SCROLL_STATE_SETTLING) {
            handleSetNextItem();
        }
        mScrollState = state;
    }

    private void handleSetNextItem() {
        final int lastPosition = mViewPager.getAdapter().getCount() - 1;
        if(mCurrentPosition == 0) {
            mViewPager.setCurrentItem(lastPosition, false);
        } else if(mCurrentPosition == lastPosition) {
            mViewPager.setCurrentItem(0, false);
        }
    }

    @Override
    public void onPageScrolled(final int position, final float positionOffset, final int positionOffsetPixels) {
    }
}

File: app/src/main/java/org/sil/storyproducer/controller/pager/CircularViewPagerHandler.kt
package org.sil.storyproducer.controller.pager

import android.support.v4.view.ViewPager
import android.util.Log

import org.sil.storyproducer.model.Workspace

/**
 * Class that implements the ViewPager.OnPageChangeListener to give the view pager circular functionality
 */
class CircularViewPagerHandler(private val mViewPager: ViewPager) : ViewPager.OnPageChangeListener {
    private var userWasScrolling = false
    private var lastScrollState = ViewPager.SCROLL_STATE_IDLE

    override fun onPageSelected(position: Int) {
        if (userWasScrolling) {
            Workspace.activeStory.lastSlideNum = position
            Log.e("@pwhite", "switched slides $position")
        }
    }

    override fun onPageScrollStateChanged(scrollState: Int) {
        if (scrollState == ViewPager.SCROLL_STATE_DRAGGING) {
            userWasScrolling = true
        } else if (scrollState == ViewPager.SCROLL_STATE_IDLE) {
            // Whether the page was scrolled to programmatically or by the
            // user, SCROLL_STATE_SETTLING always comes immediately before
            // SCROLL_STATE_IDLE. The only time it does not is when the end
            // or beginning of the ViewPager is reached, in which case, the
            // dragging occurs, but no scrolling happens, which means the scroll
            // state moves back to SCROLL_STATE_IDLE immediately
            if (lastScrollState != ViewPager.SCROLL_STATE_SETTLING) {
                val lastPosition = mViewPager.adapter!!.count - 1
                if (Workspace.activeStory.lastSlideNum == 0) {
                    mViewPager.setCurrentItem(lastPosition, false)
                } else if (Workspace.activeStory.lastSlideNum == lastPosition) {
                    mViewPager.setCurrentItem(0, false)
                }            
            }
            userWasScrolling = false
        }
        lastScrollState = scrollState
    }

    private fun handleSetNextItem() {
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
    }
}
